rls/src/main/java/io/grpc/rls/internal/LinkedHashLruCache.java

```diff
@@ -0,0 +1,381 @@
+  private final EvictionListener<K, SizedValue> evictionListener;
+  private final AtomicLong estimatedSizeBytes = new AtomicLong();
+  private long estimatedMaxSizeBytes;
+
+  LinkedHashLruCache(
+      final long estimatedMaxSizeBytes,
+      @Nullable final EvictionListener<K, V> evictionListener,
+      int cleaningInterval,
+      TimeUnit cleaningIntervalUnit,
+      ScheduledExecutorService ses,
+      final TimeProvider timeProvider) {
+    checkState(estimatedMaxSizeBytes > 0, "max estimated cache size should be positive");
+    this.estimatedMaxSizeBytes = estimatedMaxSizeBytes;
+    this.evictionListener = new SizeHandlingEvictionListener(evictionListener);
+    this.timeProvider = checkNotNull(timeProvider, "timeProvider");
+    delegate = new LinkedHashMap<K, SizedValue>(
+        // rough estimate or minimum hashmap default
+        Math.max((int) (estimatedMaxSizeBytes / 1000), 16),
+        /* loadFactor= */ 0.75f,
+        /* accessOrder= */ true) {
+      @Override
+      protected boolean removeEldestEntry(Map.Entry<K, SizedValue> eldest) {
+        if (estimatedSizeBytes.get() <= LinkedHashLruCache.this.estimatedMaxSizeBytes) {
+          return false;
+        }
+
+        // first, remove at most 1 expired entry
+        boolean removed = cleanupExpiredEntries(1, timeProvider.currentTimeNanos());
+        // handles size based eviction if necessary no expired entry
+        boolean shouldRemove =
+            !removed && shouldInvalidateEldestEntry(eldest.getKey(), eldest.getValue().value);
+        if (shouldRemove) {
+          // remove entry by us to make sure lruIterator and cache is in sync
+          LinkedHashLruCache.this.invalidate(eldest.getKey(), EvictionType.SIZE);
+        }
+        return false;
+      }
+    };
+    periodicCleaner = new PeriodicCleaner(ses, cleaningInterval, cleaningIntervalUnit).start();
+  }
+
+  /**
+   * Determines if the eldest entry should be kept or not when the cache size limit is reached. Note
+   * that LruCache is access level and the eldest is determined by access pattern.
+   */
+  @SuppressWarnings("unused")
+  protected boolean shouldInvalidateEldestEntry(K eldestKey, V eldestValue) {
+    return true;
+  }
+
+  /** Determines if the entry is already expired or not. */
+  protected abstract boolean isExpired(K key, V value, long nowNanos);
+
+  /**
+   * Returns estimated size of entry to keep track. If it always returns 1, the max size bytes
+   * behaves like max number of entry (default behavior).
+   */
+  @SuppressWarnings("unused")
+  protected int estimateSizeOf(K key, V value) {
+    return 1;
+  }
+
+  /** Updates size for given key if entry exists. It is useful if the cache value is mutated. */
+  public void updateEntrySize(K key) {
+    SizedValue entry = readInternal(key);
+    if (entry == null) {
+      return;
+    }
+    int prevSize = entry.size;
+    int newSize = estimateSizeOf(key, entry.value);
+    entry.size = newSize;
```

Comment:
i think the concurrent invocation part is fine. because of the key and value are final. so newSize is constant. 
if esimateSizeOf returns different value it can be wrong. in this case, we can't do anything to correct it. i'll add it to the doc.

Reply:
that's violating the contract that `estimateSizeOf` should be constant during the call. 
even if we do compare and swap it will still be out of sync. it should update size from `last modified size - original size` where original is before the concurrent call. CAS will ignore valid size changes unless in a while loop doing everything over. we might as well use lock  in this case, but seems a bit overkill than a good contract. 

yes, so this is not thread safe, i'll remove the thread safe annotation. hmm holding lock doesn't sound that bad than removing the guarantee, WDYT?

Reply:
I don't think that's violating the contract that `estimateSizeOf` should be constant during the call. As long as the value is not mutating during the call, it is constant.

> even if we do compare and swap it will still be out of sync
Why? If compare and swap fails we start over and retry, until it succeeds. When it succeeds it should be in sync.

Reply:
okay my bad it is not violating the exact sentence. i really meant it shouldn't be called concurrently and mistakenly use the method. but my intention was the readInternal should be identical (including size). and i also missed the retry =(. 

technically it is still not correct because of the eviction. only correct threadsafe way i can see is holding a lock. Even if we ignore the race between eviction/updateSize, lock will be faster than retry because everytime it retries it need to acquire the lock (assuming estimateSizeOf is fast - it should be).

Reply:
Using a lock looks quite simple. What is your concern for using a lock?

Reply:
yeah i liked said i think lock is good for this case. i'll make the change.